### ISSUE TITLE
Kafka topic connector - restarts on startup

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#2717]
When running the coco pharma notebook (configure, start) I see 126 shutdown events from topic connectors ie:

```
core_1       | Fri Mar 06 12:44:36 GMT 2020 cocoMDS2 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 69 unsent messages
```

The pattern is:
```
core_1       | Fri Mar 06 12:44:41 GMT 2020 cocoMDS2 Startup OCF-KAFKA-TOPIC-CONNECTOR-0002 8 properties passed to the Apache Kafka Producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic
core_1       | Fri Mar 06 12:44:41 GMT 2020 cocoMDS2 Startup OCF-KAFKA-TOPIC-CONNECTOR-0010 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is starting up with 0 buffered messages
core_1       | Fri Mar 06 12:44:42 GMT 2020 cocoMDS2 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 25 unsent messages
core_1       | Fri Mar 06 12:44:42 GMT 2020 cocoMDS2 Startup OCF-KAFKA-TOPIC-CONNECTOR-0002 8 properties passed to the Apache Kafka Producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic
core_1       | Fri Mar 06 12:44:42 GMT 2020 cocoMDS2 Startup OCF-KAFKA-TOPIC-CONNECTOR-0010 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is starting up with 0 buffered messages
```
repeated

The full output is at https://gist.github.com/a81f41773b688b63546571556d2102e4

In this case kafka is working, and the topics do exist - so something else is destabilizing the environment during startup. There is no manipulation to introduce kafka errors.

This test was done in the docker compose environment

The startup/shutdown are usually adjacent in the logs so it looks like an immediate failure of some kind.

Unsure if any messages are lost, but the environment does quicky stabilize with the connectors connected.(About 50s)

**Comment by planetf1** [comment]
Could be a regression since #2638 . I don't see this issue on builds from a few weeks back. Could also be other changes relating to framework initialization

**Comment by planetf1** [comment]
Re-running locally I see less restarting - tough there is some. 17.  May suggest a timing issue:


But looking spefically at a single server and cohort we see:
```
➜  /tmp cat out-local | grep -y shutting | grep cocoC
Fri Mar 06 13:12:10 GMT 2020 cocoMDS2 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 488 unsent messages
Fri Mar 06 13:12:10 GMT 2020 cocoMDS5 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 488 unsent messages
Fri Mar 06 13:12:10 GMT 2020 cocoMDS3 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 488 unsent messages
Fri Mar 06 13:12:10 GMT 2020 cocoMDS6 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 440 unsent messages
Fri Mar 06 13:12:10 GMT 2020 cocoMDS6 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 29 unsent messages
Fri Mar 06 13:12:10 GMT 2020 cocoMDS6 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 15 unsent messages
Fri Mar 06 13:12:13 GMT 2020 cocoMDS1 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 268 unsent messages
Fri Mar 06 13:12:13 GMT 2020 cocoMDS1 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 59 unsent messages
Fri Mar 06 13:12:13 GMT 2020 cocoMDS1 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 16 unsent messages
Fri Mar 06 13:12:14 GMT 2020 cocoMDS1 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 139 unsent messages
```
That suggests messages are 'lost' - but another possibility is that the count of messages sent isn't correct.

**Comment by planetf1** [comment]
The exception is:
```
core_1       | 2020-03-06 13:28:58.825 ERROR 1 --- [pool-2-thread-1] o.a.e.t.k.KafkaOpenMetadataEventProducer : Bad exception from sending events OCF-KAFKA-TOPIC-CONNECTOR-400-001 An unexpected java.lang.IllegalStateException exception was caught while sending an event to topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic.  The message in the exception was: Cannot perform operation after producer has been closed
core_1       | Fri Mar 06 13:28:58 GMT 2020 cocoMDS2 Shutdown OCF-KAFKA-TOPIC-CONNECTOR-0011 The Apache Kafka producer for topic egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic is shutting down after sending 1 messages and with 1 unsent messages
```
This time there were 60 -- so definately timing related

The full log with 'Info' level debugging is at https://gist.github.com/99affddff7678d4b16120dc1fe29fddf

**Comment by planetf1** [comment]
@wbittles - any ideas?

**Comment by wbittles** [comment]
@planetf1 that looks interesting assign this to me and I'll have a look , can you tar up your logs from the docker-compose  please, if there not already available.
what changes were made to the framework that may have an impact do we have an issue ?

**Comment by planetf1** [comment]
Thanks. All the logs are as gists in the text above.
I've not looked into framework changes, just aware that though I've only just noticed this today (after the new merge), the logs I'd saved were from a few weeks back.

**Comment by wbittles** [comment]
@jnhagelberg just making you aware as it's a change in cohort initialization behaviour

**Comment by planetf1** [comment]
Good idea @wbittles though I tried that PR with and without your fix, and had the same behaviour each time.

**Comment by wbittles** [comment]
@planetf1 this is quite interesting 

For referral the full trace has starts at the following timestamp
core_1       | 2020-03-06 13:28:57.742

I'll edit the entries here for readability 

The root cause for this issue  
1)The Producer connects 
2)The Producer receives the [Producer clientId=producer-1] Closing the Kafka producer with timeoutMillis = 9223372036854775807 ms. (note the nonsense value). The egeria connector is unaware that the connection has been closed. 
3)The Producer then tries to send an event, which results in the java.lang.IllegalStateException that is reported above.

The topic connector  loops until the kafka stops culling the connections, it's not obvious what else the connector can do.

**Comment by planetf1** [comment]
The value 9223372036854775807 is max for a java long, so I presume it means 'for ever' - or at least that was the intent.

So the producer has closed - but we don't know the precise reason why
If we were able to detect that we could proactively recycle the connection to avoid the Illegal state - though I agree ultimately this would offer little difference in behaviour
* It is good we recycle - and we DO recover
* I'm unsure yet why we see this now (and the corresponding audit events) but not before
* It's mostly the audit messages that could be confusing, though I'm ok we revisit when we rebuild the connector


I wonder if the cause is the relabalance. This can result in consumer groups being revoked. Maybe that's what happening here, so our producer is unassigned from it's current home.

Actually a little searching elicited this nugget : https://medium.com/streamthoughts/apache-kafka-rebalance-protocol-or-the-magic-behind-your-streams-applications-e94baf68e4f2

**Comment by planetf1** [comment]
That log message comes from https://github.com/apache/kafka/blob/937f1f741c026767b6b71ce0b0e63ae7ebe6e936/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java  line 1178

ie this is from close() with a nearly infinite timeout specified - this is so that pending messages can be sent.

So the question is who/why is the close called.

It is called if an exception occurs in the constructor - but we don't see this.

**Comment by planetf1** [comment]
Some debug logs:
```
2020-03-09 10:53:29.220 DEBUG 1 --- [ad | producer-3] org.apache.kafka.clients.Metadata        : [Producer clientId=producer-
3] Updated cluster metadata updateVersion 5 to MetadataCache{clusterId='J_Ecx75jSy-o60n3kAjdXw', nodes=[lab-kafka-0.lab-kafka
-headless.default.svc.cluster.local:9092 (id: 1001 rack: null)], partitions=[PartitionInfoAndEpoch{partitionInfo=Partition(to
pic = egeria.omag.openmetadata.repositoryservices.cohort.iotCohort.OMRSTopic, partition = 0, leader = 1001, replicas = [1001]
, isr = [1001], offlineReplicas = []), epoch=-1}], controller=lab-kafka-0.lab-kafka-headless.default.svc.cluster.local:9092 (
id: 1001 rack: null)}
2020-03-09 10:53:29.286 DEBUG 1 --- [ad | producer-3] org.apache.kafka.common.metrics.Metrics  : Added sensor with name topic
.egeria.omag.openmetadata.repositoryservices.cohort.iotCohort.OMRSTopic.records-per-batch
2020-03-09 10:53:29.287 DEBUG 1 --- [ad | producer-3] org.apache.kafka.common.metrics.Metrics  : Added sensor with name topic
.egeria.omag.openmetadata.repositoryservices.cohort.iotCohort.OMRSTopic.bytes
2020-03-09 10:53:29.287 DEBUG 1 --- [ad | producer-3] org.apache.kafka.common.metrics.Metrics  : Added sensor with name topic
.egeria.omag.openmetadata.repositoryservices.cohort.iotCohort.OMRSTopic.compression-rate
2020-03-09 10:53:29.287 DEBUG 1 --- [ad | producer-3] org.apache.kafka.common.metrics.Metrics  : Added sensor with name topic
.egeria.omag.openmetadata.repositoryservices.cohort.iotCohort.OMRSTopic.record-retries
2020-03-09 10:53:29.287 DEBUG 1 --- [ad | producer-3] org.apache.kafka.common.metrics.Metrics  : Added sensor with name topic
.egeria.omag.openmetadata.repositoryservices.cohort.iotCohort.OMRSTopic.record-errors
2020-03-09 10:53:29.287 DEBUG 1 --- [ad | producer-3] org.apache.kafka.clients.NetworkClient   : [Producer clientId=producer-3] Using older server API v7 to send PRODUCE {acks=-1,timeout=30000,partitionSizes=[egeria.omag.openmetadata.repositoryservices.cohort.iotCohort.OMRSTopic-0=1355]} with correlation id 8 to node 1001
2020-03-09 10:53:29.422  INFO 1 --- [pool-4-thread-1] o.a.k.clients.producer.KafkaProducer     : [Producer clientId=producer-3] Closing the Kafka producer with timeoutMillis = 9223372036854775807 ms.
2020-03-09 10:53:29.430 DEBUG 1 --- [ad | producer-3] o.a.k.clients.producer.internals.Sender  : [Producer clientId=producer-3] Beginning shutdown of Kafka producer I/O thread, sending remaining records.
```
Not yet able to attach/debug locally (this was k8s) as performance impact changes timing - no errors seen
(can post full log later)

**Comment by wbittles** [comment]
@planetf1 I went through all the timeouts mentioned in the gist and none of them were setting anything near that size of a timeout , plus the producer hasn't been connected anywhere near long enough to justify an exception being thrown( less than 200ms) . So I am inclined to think that that something is causing an erroneous timeout.

It's good that your seeing these log messages, the code you've been running up to now would have been stuck in an infinite loop, so progress has been made. Although that said though the logging and the lifecycle of the kafka clients are areas that do need improving.

It was far from obvious how we would check the state of the producer to enable us to avoid the illegal state exception , and all it would really achieve would be a different log message, however , that being said if I localize the lifecycle of the producer it should give me far better control of log messages and recovery.

So in summary , I think the timeout is the real cause of the problem , but I agree totally that we should have a look at making it less confusing and more robust

**Comment by wbittles** [comment]
@planetf1 Currently we create the producer during initialization, but we don't actually need it until we need to send an event, which is exactly the same window when this event occurs.
I suggest I move the connection creation from the constructor  into the first call to sendEvent. 

The first event sent on the new connector will absorb the cost of creating the connection, but considering the issues we have with cohort events this must be a good thing. It also gives us greater control over error logging and recovery.

**Comment by planetf1** [comment]
Here's a trace log of one particular producer (the full log is rather large) which gives some insight into the internals of kafka operation.

https://gist.github.com/6e9b53ce70ef76a9131f04a0d74507f7

Unfortunately the core of this log is similar to before
```
2020-03-09 15:19:21.746 TRACE 1 --- [d | producer-67] o.a.k.clients.producer.internals.Sender  : [Producer clientId=producer-67] Nodes with data ready to send: [lab-kafka-0.lab-kafka-headless.default.svc.cluster.local:9092 (id: 1001 rack: null)]
2020-03-09 15:19:21.746 DEBUG 1 --- [d | producer-67] org.apache.kafka.clients.NetworkClient   : [Producer clientId=producer-67] Using older server API v7 to send PRODUCE {acks=-1,timeout=30000,partitionSizes=[egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic-0=1358]} with correlation id 6 to node 1001
2020-03-09 15:19:21.746 TRACE 1 --- [d | producer-67] o.a.k.clients.producer.internals.Sender  : [Producer clientId=producer-67] Sent produce request to 1001: (type=ProduceRequest, acks=-1, timeout=30000, partitionRecords=({egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic-0=MemoryRecords(size=1358, buffer=java.nio.HeapByteBuffer[pos=0 lim=1358 cap=1358])}), transactionalId=''
2020-03-09 15:19:21.753 TRACE 1 --- [d | producer-67] org.apache.kafka.clients.NetworkClient   : [Producer clientId=producer-67] Completed receive from node 1001 for PRODUCE with correlation id 6, received {responses=[{topic=egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic,partition_responses=[{partition=0,error_code=0,base_offset=64,log_append_time=-1,log_start_offset=0}]}],throttle_time_ms=0}
2020-03-09 15:19:21.753 TRACE 1 --- [d | producer-67] o.a.k.clients.producer.internals.Sender  : [Producer clientId=producer-67] Received produce response from node 1001 with correlation id 6
2020-03-09 15:19:21.753 TRACE 1 --- [d | producer-67] o.a.k.c.p.internals.ProducerBatch        : Successfully produced messages to egeria.omag.openmetadata.repositoryservices.cohort.cocoCohort.OMRSTopic-0 with base offset 64.
2020-03-09 15:19:21.753  INFO 1 --- [pool-2-thread-1] o.a.k.clients.producer.KafkaProducer     : [Producer clientId=producer-67] Closing the Kafka producer with timeoutMillis = 9223372036854775807 ms.
2020-03-09 15:19:21.753 DEBUG 1 --- [d | producer-67] o.a.k.clients.producer.internals.Sender  : [Producer clientId=producer-67] Beginning shutdown of Kafka producer I/O thread, sending remaining records
```
ie things are looking ok, then somehow close() is called.

**Comment by planetf1** [comment]
Not managed to find where the close may get called directly in the kafka code -- the last 'ProducerBatch' went well. Nor can I see the close from our code. However in kafka itself there is a lot of async activity so easy to miss. Have been unable to reproduce with debugging enabled (so I could break on the call to close() )

Your suggestion @wbittle may change the dynamics  and avoid the issue, though I still don't think we know why.

That being said, importantly the connector does recover into a good working state - and as long as in future we can ensure we're not losing messages, that's about as good as we can be for now.

Welcome any input from @grahamwallis @mandy-chessell but it may be appropriate to close this as wontfix - at least if someone sees this issue they can find it in github.

**Comment by wbittles** [comment]
@planetf1 what's really confusing is that there is no log entries for egeria actually calling into kafka to try and send the event, that being said perhaps that producer.close I put in the finally block looks like the prime suspect. I'm building the fix now

**Comment by wbittles** [comment]
I've just opened a PR that moves the producer lifecycle into the KafkaOpenMetadataEventProducer.publishEvent method. It means we only have kafka connection when we actually need them, and only specifically close them on receiving an exception

**Comment by wbittles** [comment]
As for the concern around events being processed correctly , I could change the producers buffer to have a max size (via a property) which would block events being produced when the limit was reached , until kafka recovered (or failed) . This would allow us to control the production of events from totally blocking to highly asynchronous. The blocking would really aid in overriding bringup timing issues.

**Comment by planetf1** [comment]
Were you thinking of the buffer size? We could set this externally without code changes I think.  We could also try max.in.flight.requests.per.connection = 1 to limit overlaps (for a test) . Also if we normally expect behaviour to be async, we need to figure out how to debug that scenario. 

For the PR - I added a xref and made code comments there.

**Comment by wbittles** [comment]
@planetf1 The Egeria Connector has two internal buffers , one for producers and one for consumers. Currently the producer buffer is an unlimited list . I was thinking of changing that implementation to one that uses a blocking queue with a configurable fixed size. It would give the ability to process each event sequencely if you wished, while protecting from buffer overflow type problems. 

as for default configuration changes for the producer we absolutely need to revisit most of our default setting , we do send synchronously by waiting on the future so no real need to worry about debugging that case

**Comment by jnhagelberg** [comment]
@wbittles Would it be possible to set the size to -1 if we don't want a fixed size?  That seems better than configuring the size to Integer.MAX_VALUE, which might end up creating a giant list up front.

**Comment by planetf1** [comment]
The problem seems to be addressed. @wbittles if you want to propose any followups it's probably best we open up a new issue or you add notes to an existing issue? Any general documentation points may be best addressed via #2947
